rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,628 @@
+                    Ok((_rem, suback)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::SUBACK(suback),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                10 => match parse_unsubscribe(rem, protocol_version) {
+                    Ok((_rem, unsub)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::UNSUBSCRIBE(unsub),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                11 => match parse_unsuback(rem, protocol_version) {
+                    Ok((_rem, unsuback)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::UNSUBACK(unsuback),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                12..=13 => {
+                    let msg = MQTTMessage {
+                        header: header,
+                        op: match message_type {
+                            12 => MQTTOperation::PINGREQ,
+                            13 => MQTTOperation::PINGRESP,
+                            _ => MQTTOperation::UNKNOWN, // should not happen
+                        },
+                    };
+                    if len > rem.len() {
+                        return Err(Err::Incomplete(Needed::Size(len - rem.len())));
+                    }
+                    return Ok((&rem[len..], msg));
+                }
+                14 => match parse_disconnect(rem, len, protocol_version) {
+                    Ok((_rem, disco)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::DISCONNECT(disco),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                15 => match parse_auth(rem) {
+                    Ok((_rem, auth)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::AUTH(auth),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                _ => {
+                    let msg = MQTTMessage {
+                        header: header,
+                        op: MQTTOperation::UNKNOWN, // should not happen
+                    };
```

Comment:
can you panic when in fuzz mode if this is not supposed to happen ?

Reply:
Let me rephrase: It is still possible for this case to be triggered with broken traffic (message type set to reserved value 0). I wouldn't panic here since this does not indicate a programming error but rather an anomaly in traffic. One might, for example, raise an event ("use of reserved message type 0") here. Ok, I can add that.

Reply:
Ok for the event if it is not covered by another signature like `mqtt.type:0` .
I would modify the comment to say that this should not happen with normal traffic.
The other comments `should not happen` you made seem unreachable code

Reply:
Okay, right.

Reply:
FYI, we're adding `debug_validate_bug_on` which will operate the same as `DEBUG_VALIDATE_BUG_ON` in our C code.

But it sounds like it's not appropriate here.